aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -443,7 +431,12 @@ _embedding_bag_cpu(const Tensor &weight, const Tensor &indices,
     TORCH_CHECK(per_sample_weights.numel() == indices.numel());
   }
 
-  auto bag_size = make_bag_size(offsets, indices, mode, weight.requires_grad());
+  Tensor bag_size;
+  if (requires_grad) {
```

Comment:
I feel we're adding too many layers here. 
We should move `requires_grad` up before dispatching to cpu/cuda kernels since this logic is device agnostic. 
Something I'm thinking about is 
```
embedding_bag
  # handles requires_grad here
  if (requires_grad)
    _embedding_bag_forward_only
  else
    _embedding_bag

# in native_functions.yaml
set device dispatch for _embedding_bag_forward_only and _embedding_bag, 
remove device dispatch for embedding_bag
```
^^^ This is just psuedocode, names are temporary that you don't need to follow. I mainly want to mention the logic should be above device dispatch.

Reply:
The referenced code (branching on requires_grad) is purely for code sharing. They are not exactly two ops. Otherwise, _embedding_bag_forward_only and _embedding_bag would have almost identical implementation. Let me know what you think.

The other thing is the level of abstraction to actually branch on weight.requires_grad. Current implementation is at python level. What you are proposing is inside implementation of embedding_bag's kernel (even if it just calls another op). I thought our goal is to remove reading of weight.requires_grad in kernels. Am I missing something?

Reply:
Discussed offline. This _embedding_bag_cpu_impl is just a helper function in aten::native rather than a exposed aten op. 

Moved branching logic to top-level op embedding_bag.